File: SmartLink/Source/SmartLinkClient/NRC_ServiceGenericClasses/TaskSchedulerClasses/TaskScheduler_Scheduler.cs
using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.IO;
using System.Diagnostics;
//********************************************************************'
// Created by Elibad - 09/01/2007
//
//  This class is the base class for the Task Schedulers
//  These Classes can schedule any type of task based 
//  on different interval types (Hourly, Daily, Weekly, etc)
//*********************************************************************

namespace NRC.Miscellaneous.TaskScheduler
{
	public abstract class Scheduler : IDisposable
	{

		private int _iInterval;
		private DateTime _dtLastRun;
		private DateTime _dtNextRun;
		private bool _bEnabled;
		private System.DateTime _dStartDate;
		private DateTime _dtTime;

		private System.Timers.Timer _tTimer = new System.Timers.Timer();
		protected bool disposed = false;

		protected bool _bFirstRun;
		/// <summary>
		/// Generates the Next Run Date
		/// </summary>
		protected abstract DateTime GetNextOccurrence(DateTime BaseDate);

		/// <summary>
		/// This Event gets Trigered when the schedule reaches the next run Date and time
		/// </summary>
		public event RunTaskEventHandler RunTask;
		public delegate void RunTaskEventHandler();

		/// <summary>
		/// Interval of recurrence
		/// </summary>
		/// <remarks>
		/// Examples:
		/// Interval = 2
		/// 
		/// If Recurrence Type = Daily then it will run every 2 Days
		/// If Recurrence Type = Weekly then it will run every 2 Weeks
		/// </remarks>
		public int Interval {
			get { return _iInterval; }
			set {
				if (value > 0) {
					_iInterval = value;
					if (this.Enabled) {
						this._bFirstRun = true;
						this.UpdateNextOccurrence();
					}
				} else {
					throw new System.Exception("Incorrect Interval value" + Environment.NewLine + Environment.NewLine + "Interval should be higher than 0");
				}
			}
		}

		/// <summary>
		/// Date to start schedule
		/// </summary>
		/// <value>Default value is Date.Now</value>
		public string StartDate {
			get { return _dStartDate.ToString("MM/dd/yyyy"); }
			set {
				try {
					_dStartDate = System.DateTime.Parse(value);
				} catch {
					throw new System.Exception("Incorrect Date" + Environment.NewLine + Environment.NewLine + "Please provide a date using mm/dd/yyyy format");
				}

				if (this.Enabled) {
					this.UpdateNextOccurrence();
				}
			}
		}

		/// <summary>
		/// Defines if the schedule is running
		/// </summary>
		public bool Enabled {
			get { return _bEnabled; }
			set {
				_bEnabled = value;
				_tTimer.Enabled = value;
				if (this.Enabled) {
					this._bFirstRun = true;
					this.UpdateNextOccurrence();
				}
			}
		}


		/// <summary>
		/// Time of the Day for Next recurrence
		/// </summary>
		public string Time {
			get { return _dtTime.ToString("HH:mm"); }
			set {
				try {
					_dtTime = DateTime.Parse(value);
				} catch  {
					throw new System.Exception("Incorrect Time" + Environment.NewLine + Environment.NewLine + "Please provide correct time value (24 hh:mm or hh:mm AM/PM)");
				}

				if (this.Enabled) {
					this._bFirstRun = true;
					this.UpdateNextOccurrence();
				}
			}
		}

		public Scheduler() : this("01:00:00")
		{
		}

		public Scheduler(string Time) : this(1, System.DateTime.Today.ToString("MM/dd/yyyy"), Time, false)
		{
		}

		public Scheduler(int Interval, string StartDate, string Time, bool Enabled)
		{
			//Initialize variables
			this.Interval = Interval;
			this.StartDate = StartDate;
			this.Time = Time;
			this._bEnabled = Enabled;
			this._bFirstRun = true;

			_dtLastRun = System.DateTime.Now;
			_dtNextRun = System.DateTime.Now.AddDays(3);

			//Initialize internal Timer
			_tTimer.Elapsed += VerifySchedule;

			// Set the timer interval to the number of minutes set in the app.config file
			_tTimer.Interval = new TimeSpan(0, 0, 10).TotalMilliseconds;

			_tTimer.AutoReset = false;

			// Turn on _tTimer
			_tTimer.Enabled = Enabled;
		}

		/// <summary>
		/// This Method is executed by the internal timer Tick event
		/// </summary>
		/// <remarks>This Method verifies wheter the Next run date has been reached</remarks>
		private void VerifySchedule(object sender, System.Timers.ElapsedEventArgs e)
		{
			if (DateAndTime.Now > _dtNextRun) {
				_dtLastRun = _dtNextRun;
				this.UpdateNextOccurrence();
				if (RunTask != null) {
					RunTask();
				}
			}
			_tTimer.Start();
		}

		/// <summary>
		/// This Method updates the Next Run Date and Time with the proper value
		/// </summary>
		protected void UpdateNextOccurrence()
		{
			if (_dtLastRun > _dStartDate) {
				_dtNextRun = GetNextOccurrence(_dtLastRun);
			} else if (_dStartDate > System.DateTime.Now) {
				_dtNextRun = GetNextOccurrence(_dStartDate);
			}
		}
		public void Start()
		{
			this.Enabled = true;
		}

		public void Stop()
		{
			this.Enabled = false;
		}

		/// <summary>
		/// This value determines how often the scheduler will verify that the next scheduled run has been reached
		/// </summary>
		/// <value>Default = 30 Seconds</value>
		public int ScheduleCheckInterval {
			get { return Convert.ToInt32(_tTimer.Interval / 1000); }
			set {
				if (value > 0) {
					_tTimer.Interval = value * 1000;
				}
			}
		}

		#region "IDisposable support"
		protected virtual void dispose(bool disposing)
		{
			if (!this.disposed) {

				if (disposing) {
				}
				//SaveRunDates()
			}
			this.disposed = true;
		}

		public void Dispose()
		{
			dispose(true);
			GC.SuppressFinalize(this);
		}

        //protected override void Finalize()
        //{
        //    dispose(false);
        //    base.Finalize();
        //}

		/// <summary>
		/// Gets the Date and Time for Next Run
		/// </summary>
		public string NextRun {
			get {
				string functionReturnValue = null;
				functionReturnValue = _dtNextRun.ToString("MM/dd/yyyy HH:mm");
				if (this.Enabled == false) {
					functionReturnValue = "Scheduler is Disabled";
				}
				return functionReturnValue;
			}
		}

		#endregion

	}

	//Public MustInherit Class NRC_Scheduler
}

File: SmartLink/Source/SmartLinkClient/NRC_ServiceGenericClasses/TaskSchedulerClasses/TaskScheduler_Scheduler.vb
Option Explicit On
'********************************************************************'
' Created by Elibad - 09/01/2007
'
'  This class is the base class for the Task Schedulers
'  These Classes can schedule any type of task based 
'  on different interval types (Hourly, Daily, Weekly, etc)
'*********************************************************************

Namespace Miscellaneous.TaskScheduler
    Public MustInherit Class Scheduler
        Implements IDisposable

        Private _iInterval As Integer
        Private _sFileName As String
        Private _dtLastRun As DateTime
        Private _dtNextRun As DateTime
        Private _bEnabled As Boolean
        Private _dStartDate As Date
        Private _dtTime As DateTime
        Private _tTimer As New System.Timers.Timer

        Protected disposed As Boolean = False
        Protected _bFirstRun As Boolean

        ''' <summary>
        ''' Generates the Next Run Date
        ''' </summary>
        Protected MustOverride Function GetNextOccurrence(ByVal BaseDate As DateTime) As DateTime

        ''' <summary>
        ''' This Event gets Trigered when the schedule reaches the next run Date and time
        ''' </summary>
        Public Event RunTask()

        ''' <summary>
        ''' Interval of recurrence
        ''' </summary>
        ''' <remarks>
        ''' Examples:
        ''' Interval = 2
        ''' 
        ''' If Recurrence Type = Daily then it will run every 2 Days
        ''' If Recurrence Type = Weekly then it will run every 2 Weeks
        ''' </remarks>
        Public Property Interval() As Integer
            Get
                Interval = _iInterval
            End Get
            Set(ByVal value As Integer)
                If value > 0 Then
                    _iInterval = value
                    If Me.Enabled Then
                        Me._bFirstRun = True
                        Me.UpdateNextOccurrence()
                    End If
                Else
                    Throw New System.Exception("Incorrect Interval value" & vbCrLf & vbCrLf & "Interval should be higher than 0")
                End If
            End Set
        End Property

        ''' <summary>
        ''' Date to start schedule
        ''' </summary>
        ''' <value>Default value is Date.Now</value>
        Public Property StartDate() As String
            Get
                StartDate = _dStartDate.ToString("MM/dd/yyyy")
            End Get
            Set(ByVal value As String)
                Try
                    _dStartDate = Date.Parse(value)
                Catch ex As Exception
                    Throw New System.Exception("Incorrect Date" & vbCrLf & vbCrLf & "Please provide a date using mm/dd/yyyy format")
                End Try

                If Me.Enabled Then
                    Me.UpdateNextOccurrence()
                End If
            End Set
        End Property

        ''' <summary>
        ''' Defines if the schedule is running
        ''' </summary>
        Public Property Enabled() As Boolean
            Get
                Enabled = _bEnabled
            End Get
            Set(ByVal value As Boolean)
                _bEnabled = value
                _tTimer.Enabled = value
                If Me.Enabled Then
                    Me._bFirstRun = True
                    Me.UpdateNextOccurrence()
                End If
            End Set
        End Property


        ''' <summary>
        ''' Time of the Day for Next recurrence
        ''' </summary>
        Public Property Time() As String
            Get
                Time = _dtTime.ToString("HH:mm")
            End Get
            Set(ByVal value As String)
                Try
                    _dtTime = DateTime.Parse(value)
                Catch ex As Exception
                    Throw New System.Exception("Incorrect Time" & vbCrLf & vbCrLf & "Please provide correct time value (24 hh:mm or hh:mm AM/PM)")
                End Try

                If Me.Enabled Then
                    Me._bFirstRun = True
                    Me.UpdateNextOccurrence()
                End If
            End Set
        End Property

        Public Sub New()
            Me.New("01:00:00")
        End Sub

        Public Sub New(ByVal Time As String)
            Me.New(1, Date.Today.ToString("MM/dd/yyyy"), Time, False)
        End Sub

        Public Sub New(ByVal Interval As Integer, ByVal StartDate As String, ByVal Time As String, ByVal Enabled As Boolean)
            'Initialize variables
            Me.Interval = Interval
            Me.StartDate = StartDate
            Me.Time = Time
            Me._bEnabled = Enabled
            Me._bFirstRun = True

            If Not Security.ValidateKey Then
                Throw New System.Exception("Not valid Security Key, Please provide a valid Security Key using the security class")
            End If

            _dtLastRun = Date.Now
            _dtNextRun = Date.Now.AddDays(3)

            'Initialize internal Timer
            AddHandler _tTimer.Elapsed, AddressOf VerifySchedule

            ' Set the timer interval to the number of minutes set in the app.config file
            _tTimer.Interval = New TimeSpan(0, 0, 10).TotalMilliseconds

            _tTimer.AutoReset = False

            ' Turn on _tTimer
            _tTimer.Enabled = Enabled
        End Sub

        ''' <summary>
        ''' This Method is executed by the internal timer Tick event
        ''' </summary>
        ''' <remarks>This Method verifies wheter the Next run date has been reached</remarks>
        Private Sub VerifySchedule(ByVal sender As Object, ByVal e As System.Timers.ElapsedEventArgs)
            If Now > _dtNextRun Then
                _dtLastRun = _dtNextRun
                Me.UpdateNextOccurrence()
                RaiseEvent RunTask()
            End If
            _tTimer.Start()
        End Sub

        ''' <summary>
        ''' 
        ''' </summary>
        Private Sub SaveRunDates()
            Dim sDates As String
            sDates = "LastRun = '" & _dtLastRun.ToString("MM/dd/yyyy HH:mm:ss") & "'"
            '            & vbCrLf & "NextRun = '" & _dtNextRun.ToString("MM/dd/yyyy HH:mm:ss") & "'"
            My.Computer.FileSystem.WriteAllText(_sFileName, sDates, False)
        End Sub

        ''' <summary>
        ''' 
        ''' </summary>
        Private Sub LoadRunDates()
            Dim sData As String
            Dim iStart As Integer
            Dim iLenght As Integer

            If System.IO.File.Exists(_sFileName) Then
                sData = My.Computer.FileSystem.ReadAllText(_sFileName)
                iStart = "LastRun = '".Length + 1
                iLenght = sData.IndexOf("'", iStart) - (iStart)
                sData = "#" & sData.Substring(iStart, iLenght) & "#"
                _dtLastRun = Date.Parse(sData)
            End If
            If _dtLastRun < Date.Now.AddYears(-1) Then
                _dtLastRun = Date.Now
            End If
            If Me.Enabled Then
                Me.UpdateNextOccurrence()
            End If
        End Sub

        ''' <summary>
        ''' This Method updates the Next Run Date and Time with the proper value
        ''' </summary>
        Protected Sub UpdateNextOccurrence()
            If _dtLastRun > _dStartDate Then
                _dtNextRun = GetNextOccurrence(_dtLastRun)
            ElseIf _dStartDate > Date.Now Then
                _dtNextRun = GetNextOccurrence(_dStartDate)
            End If
        End Sub
        Public Sub Start()
            Me.Enabled = True
        End Sub

        Public Sub [Stop]()
            Me.Enabled = False
        End Sub

        ''' <summary>
        ''' This value determines how often the scheduler will verify that the next scheduled run has been reached
        ''' </summary>
        ''' <value>Default = 30 Seconds</value>
        Public Property ScheduleCheckInterval() As Integer
            Get
                ScheduleCheckInterval = CInt(_tTimer.Interval / 1000)
            End Get
            Set(ByVal value As Integer)
                If value > 0 Then
                    _tTimer.Interval = value * 1000
                End If
            End Set
        End Property

#Region "IDisposable support"
        Protected Overridable Sub dispose(ByVal disposing As Boolean)
            If Not Me.disposed Then
                If disposing Then

                End If
                'SaveRunDates()
            End If
            Me.disposed = True
        End Sub

        Public Sub Dispose() Implements IDisposable.Dispose
            dispose(True)
            GC.SuppressFinalize(Me)
        End Sub

        Protected Overrides Sub Finalize()
            dispose(False)
            MyBase.Finalize()
        End Sub

        ''' <summary>
        ''' Gets the Date and Time for Next Run
        ''' </summary>
        Public ReadOnly Property NextRun() As String
            Get
                NextRun = _dtNextRun.ToString("MM/dd/yyyy HH:mm")
                If Me.Enabled = False Then
                    NextRun = "Scheduler is Disabled"
                End If
            End Get
        End Property

#End Region

    End Class

    'Public MustInherit Class NRC_Scheduler
End Namespace
